src/GraphQL/Extensions/GraphQLBuilderExtensions.cs

```diff
@@ -1168,5 +1171,52 @@ public static IGraphQLBuilder AddExecutionStrategy<TExecutionStrategy>(this IGra
             return builder;
         }
         #endregion
+
+        #region - UseTelemetry -
+#if NET5_0_OR_GREATER
+        /// <summary>
+        /// Configures the GraphQL server to collect traces via the <see cref="System.Diagnostics.Activity">System.Diagnostics.Activity API</see> and records events that match the
+        /// <see href="https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/graphql.md">OpenTelemetry recommendations</see>.
+        /// Trace data contains the GraphQL operation name, the operation type, and the optionally the document.
+        /// </summary>
+        /// <remarks>
+        /// When applicable, place after calls to UseAutomaticPersistedQueries to ensure that the query document is recorded properly.
+        /// </remarks>
+        public static IGraphQLBuilder UseTelemetry(this IGraphQLBuilder builder, ActivitySource activitySource, Action<TelemetryOptions>? configure = null)
+            => UseTelemetry(builder, _ => activitySource, configure != null ? (opts, _) => configure(opts) : null);
+
+        /// <inheritdoc cref="UseTelemetry(IGraphQLBuilder, ActivitySource, Action{TelemetryOptions}?)"/>
+        public static IGraphQLBuilder UseTelemetry(this IGraphQLBuilder builder, Func<IServiceProvider, ActivitySource> activitySourceFactory, Action<TelemetryOptions, IServiceProvider>? configure = null)
+        {
+            if (activitySourceFactory == null)
+                throw new ArgumentNullException(nameof(activitySourceFactory));
+
+            builder.Services.Configure(configure);
+            builder.ConfigureExecution(async (options, next) =>
+            {
+                var activitySource = activitySourceFactory(options.RequestServices ?? throw new MissingRequestServicesException());
+                using var activity = activitySource.StartActivity("graphql");
+                if (activity == null) // if no event listeners
+                    return await next(options).ConfigureAwait(false);
+                // record the requested operation name and optionally the GraphQL document
```

Comment:
It's recommended to check `activity.IsAllDataRequested` before populating tags to avoid unnecessary locations. If I understand this property correctly it can be `false` when there are active listeners but the trace was skipped by the sampler.

Reply:
So over here https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/README.md#basic-usage it says:

> It is highly recommended to check activity.IsAllDataRequested, before populating any tags which are not readily available

In our case, `options.OperationName` and `options.Query` are both already `string` values in memory, and as such are 'readily available'.  Memory consumption will not change since the query is required to stay in memory during the entire execution.  The tag names are also constants (`"graphql.operation.name"` and similar) and as such are already [interned](https://learn.microsoft.com/en-us/dotnet/api/system.string.intern?view=net-7.0).

So, based on the available documentation, should we set the tags?  Are they used at all?  If so, might as well have all of them.  It would seem that if the event was not going to be logged, `StartActivity` would have returned `null` in the first place.  What about considering the name of the property `IsAllDataRequested` -- the implication would seem that if the value is `false`, still "some" data may be requested.  What data?  Perhaps the display name even if no tags?

As a side note, I would like this PR to be generic for any telemetry writer based on System.Diagnostic.Activity.  So I would prefer not to add conditions specific to OpenTelemetry.

Your thoughts?

Reply:
Note that if we add the ability to record variables or extensions, this may require serializing the data and hence should certainly respect the `IsAllDataRequested` property.  (Or we simply pass the object in without serializing it, in which case the data is "readily available".  Presumably the OpenTelemetry API would serialize the data when saving it or something.)

Reply:
`IsAllDataRequested` is equivalent to the [`IsRecording`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#isrecording). If it's `false` all tags will be discarded anyway. For example, Asp.Net instrumentation can configure a filter and decide the request shouldn't create the trace. In fact, the trace is created, but the `tracestate` indicates it shouldn't be recorded. It's done to allow the decision propagation out of the process, so the following process won't make the opposite decision and create the trace.

> As a side note, I would like this PR to be generic for any telemetry writer based on System.Diagnostic.Activity. So I would prefer not to add conditions specific to OpenTelemetry.
The `System.Diagnostic.Activity` implements OpenTelemetry specifications, so it will always behave according to it.

Reply:
> Presumably the OpenTelemetry API would serialize the data when saving it or something.)

I think it just calls ToString() but need to verify it

Reply:
>  What about considering the name of the property IsAllDataRequested -- the implication would seem that if the value is false, still "some" data may be requested. What data? Perhaps the display name even if no tags?

Agree. The name `IsAllDataRequested` sucks.

> It's recommended to check activity.IsAllDataRequested before populating tags to avoid unnecessary allocations. 

I tend to agree with @Shane32 . Let's avoid `IsAllDataRequested` calls - at least for cases when we already have tag names and values as is ready to be written inside.

Reply:
> I think it just calls ToString() but need to verify it

It depends on exporter. Let's take two. `ConsoleExporter` works very different from `OtlpTraceExporter`. Both exporters use _tag transformers_ - `ConsoleTagTransformer` or  `OtlpKeyValueTransformer`. The first always converts tag value to string . The second - to `OtlpCommon.AnyValue` (Protobuf internals):
```cs
 /// <summary>
  /// AnyValue is used to represent any type of attribute value. AnyValue may contain a
  /// primitive value such as a string or integer or it may contain an arbitrary nested
  /// object containing arrays, key-value lists and primitives.
  /// </summary>
  internal sealed partial class AnyValue : pb::IMessage<AnyValue>
```

Reply:
So for now verdict is to avoid `IsAllDataRequested`. We can always add it later if required without any breaking changes.

Reply:
> So for now verdict is to avoid `IsAllDataRequested`.

Changed -- now that we are providing a sanitize method and according to .NET spec, it should be checked if there is any substantial work done to record the tag.  Sanitizing a document would certainly qualify.

https://learn.microsoft.com/en-us/dotnet/core/diagnostics/distributed-tracing-instrumentation-walkthroughs

> If you are instrumenting functions with high-performance requirements, [Activity.IsAllDataRequested](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.isalldatarequested#system-diagnostics-activity-isalldatarequested) is a hint that indicates whether any of the code listening to Activities intends to read auxiliary information such as Tags. If no listener will read it, then there is no need for the instrumented code to spend CPU cycles populating it.

Reply:
I agree.